### DISCUSSION TITLE
Speedup build

### DISCUSSION THREAD

**Discussion by thewh1teagle** [k2-fsa/sherpa-onnx#1215] [Q&A]
It takes me nearly 5 minutes to build it on Windows 11 (amd ryzen 5 4500u / 40mbps network speed)
Maybe it's due to slow network? not sure.
And chance we can improve it? Maybe add more feature flags similar to `tts` to make it possible to build only small portion of the library?

Also, what's the simplest way to run test with the library? maybe something that doesn't require to download model? or some small model?

**Comment by csukuangfj** [comment]
> It takes me nearly 5 minutes to build it on Windows 11 (amd ryzen 5 4500u / 40mbps network speed)

Could you describe what commands you are using?

**Reply by thewh1teagle** [reply]
> Could you describe what commands you are using?

I build it from Rust build script which invoke Cmake. I configure the Cmake args inside this file [build.rs](https://github.com/thewh1teagle/sherpa-rs/blob/main/sys/build.rs). It's similar to passing it as command line options.

*Verbose build*:
https://github.com/thewh1teagle/sherpa-rs/actions/runs/10237253821/job/28320330272#step:7:110

*Normal build*:
https://github.com/thewh1teagle/sherpa-rs/actions/runs/10237035753/job/28319833261#step:7:13

*Cmake variables*:
https://github.com/thewh1teagle/sherpa-rs/blob/main/sys/build.rs#L203-L209

*Cmake build command*:
https://github.com/thewh1teagle/sherpa-rs/blob/main/sys/build.rs#L238-L241

**Comment by csukuangfj** [comment]
Could you follow
https://k2-fsa.github.io/sherpa/onnx/install/windows.html
to build sherpa-onnx manually on your computer instead of invoking `sherpa-rs` to build it.

If it still takes 5 minutes, we can have a look.
If not, please compare your build command with the one in the sherpa-onnx's documentation.

**Reply by thewh1teagle** [reply]
Yes it takes near 5 minutes (same) when building it directly.

*Build it with:*

```console
git clone https://github.com/k2-fsa/sherpa-onnx
cd sherpa-onnx
cmake -B build . -DCMAKE_BUILD_TYPE=Release -DSHERPA_ONNX_ENABLE_BINARY=ON 
cmake --build build --config Release -j8

```

*OS: Windows 11*
*CPU: amd ryzen 5 4500u*

**Reply by csukuangfj** [reply]
Since you have run two commands, could you measure the time required for each command?

Also, it is not correct to use `-j8` on windows.

Please use
```
cmake --build build --config Release -- -m:2
```
to use `2` jobs and use `-m:5` to use 5 jobs.

**Reply by thewh1teagle** [reply]
*Config*
```console
Measure-Command { cmake -B build . -DCMAKE_BUILD_TYPE=Release -DSHERPA_ONNX_ENABLE_BINARY=ON } 
TotalMinutes      : 1.83
```

*Build*
```console
Measure-Command { cmake --build build --config Release -- /m:6 }
TotalMinutes      : 4.04
```

**Reply by csukuangfj** [reply]
How many cpus do you have?

**Reply by thewh1teagle** [reply]
> How many cpus do you have?

I have one cpu with 6 cores

```console
(Get-WmiObject -Class Win32_ComputerSystem).NumberOfLogicalProcessors
6
```

**Reply by csukuangfj** [reply]
Could you try
```
cmake -B build . \
  -DCMAKE_BUILD_TYPE=Release \
  -DSHERPA_ONNX_ENABLE_BINARY=OFF \
  -DSHERPA_ONNX_ENABLE_WEBSOCKET=OFF \
  -DSHERPA_ONNX_BUILD_C_API_EXAMPLES=OFF
```

**Reply by thewh1teagle** [reply]
*Config*
```console
 Measure-Command { cmake -B build . -DCMAKE_BUILD_TYPE=Release -DSHERPA_ONNX_ENABLE_BINARY=ON -DSHERPA_ONNX_ENABLE_WEBSOCKET=OFF -DSHERPA_ONNX_BUILD_C_API_EXAMPLES=OFF }
TotalMinutes      : 1.88
```

*Build*
```console
Measure-Command { cmake --build build --config Release -- /m:6 }
TotalMinutes      : 3.60
```

Looks a bit faster

**Reply by csukuangfj** [reply]
Please set `-DSHERPA_ONNX_ENABLE_BINARY=OFF`

**Reply by csukuangfj** [reply]
Also, if you care about the building time, please consider using our released library files.

**Reply by thewh1teagle** [reply]
*Config*
```console
Measure-Command { cmake -B build . -DCMAKE_BUILD_TYPE=Release -DSHERPA_ONNX_ENABLE_BINARY=OFF -DSHERPA_ONNX_ENABLE_WEBSOCKET=OFF -DSHERPA_ONNX_BUILD_C_API_EXAMPLES=OFF  }
TotalMinutes      : 1.57
```

*Build*

```console
Measure-Command { cmake --build build --config Release -- /m:6 }
TotalMinutes      : 3.65
```

Almost 6 minutes.


macOS m1

*Config*
```console
time cmake -B build . -DCMAKE_BUILD_TYPE=Release -DSHERPA_ONNX_ENABLE_BINARY=OFF -DSHERPA_ONNX_ENABLE_WEBSOCKET=OFF -DSHERPA_ONNX_BUILD_C_API_EXAMPLES=OFF -DCMAKE_BUILD_PARALLEL_LEVEL=8
0.82s user 5.48s system 28% cpu 57.279 total
```

*Build*
```console
time cmake --build build --config Release
121.02s user 8.53s system 98% cpu 2:11.06 total
```
